### PR TITLE
Update lesson extras UI to use style similar to bubble choice levels

### PR DESCRIPTION
**Pull request by dmcavoy**
Hack Days Project: Improve Lesson Extras UI

Improve the way we show the challenges for lesson extras to be more consistent with the way we show the challenges for bubble choice.  This change impacts Bubble Choice Levels as well as Lesson Extras Levels.

In addition while I was in this part of the code base I continued the work to rename stage to lesson.

## Bubble Choice

| Before | After |
| --- | --- |
| Screen Shot 2020-04-28 at 2 34 45 PM |  Screen Shot 2020-04-28 at 2 26 45 PM |

## Lesson Extras

| Before | After |
| --- | --- |
| Screen Shot 2020-04-28 at 2 34 30 PM |Screen Shot 2020-04-28 at 2 25 52 PM |

## Testing story

Updated test to match updates to components.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by davidsbailey** [comment]
I like the improvement to only have one orange "call to action" button on the extras page! The last time I was in a classroom (with Amanda), the lesson extras page was a major stumbling block for the elementary school students, causing the teacher to have to go around to each of their computers to help them proceed, especially because not all of them could read. This could at least make it easier to say "click the orange button", which a pre-reader can do.

File: apps/test/unit/code-studio/components/SublevelCardTest.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  sublevel: {
+    id: 1,
+    display_name: 'Choice 1',
+    description: 'Sublevel 1 is lots of fun',
+    thumbnail_url: 'some-fake.url/kittens.png',
+    url: '/s/script/stage/1/puzzle/2/sublevel/1',
+    position: 1,
+    letter: 'a',
+    perfect: false,
+    status: 'perfect'
```
Comment (Erin007) [inline apps/test/unit/code-studio/components/SublevelCardTest.js:17]: I'm a little confused about why perfect could be false, but status is 'perfect'?

Comment (dmcavoy) [reply apps/test/unit/code-studio/components/SublevelCardTest.js:17]: lol i can't. Thats just bad test data. I will update it. Thanks for catching

File: dashboard/app/models/script_level.rb [hunk 2]
```diff
@@ -361,19 +361,15 @@ def self.summarize_extra_puzzle_pages(last_level_summary)
...
-      type: level.type,
-      map: JSON.parse(level.try(:maze) || '[]'),
-      serialized_maze: level.try(:serialized_maze) && JSON.parse(level.try(:serialized_maze)),
-      skin: level.try(:skin),
-      thumbnail_url: level.try(:thumbnail_url),
-      solution_image_url: level.try(:solution_image_url),
-      level: level.summarize_as_bonus.camelize_keys,
-    }.camelize_keys
+      description: level.try(:bubble_choice_description),
+      display_name: level.display_name || 'Bonus Level',
```
Comment (Erin007) [inline dashboard/app/models/script_level.rb]: does this need to be translatable?

File: dashboard/test/models/script_test.rb [hunk 3]
```diff
@@ -1308,8 +1309,8 @@ class SummarizeVisibleAfterScriptTests < ActiveSupport::TestCase
     create :script_level, script: script, lesson: lesson3, bonus: true
     create :script_level, script: script, lesson: lesson3, bonus: true
 
-    bonus_levels1 = script.get_bonus_script_levels(lesson1)
-    bonus_levels3 = script.get_bonus_script_levels(lesson3)
+    bonus_levels1 = script.get_bonus_script_levels(lesson1, student)
+    bonus_levels3 = script.get_bonus_script_levels(lesson3, student)
```
Comment (Erin007) [inline dashboard/test/models/script_test.rb:1313]: the method takes a user_id as a param, but you're passing a whole user here; does it need to be student.id?

Comment (dmcavoy) [reply dashboard/test/models/script_test.rb:1313]: get_bonus_script_levels takes the whole user object -> summarize_as_bonus just takes the user id

**Comment by dmcavoy** [comment]
Something went wrong in pulling the rename out.... please hold on reviewing

**Comment by dmcavoy** [comment]
Alright all set. Issue has been resolved. Please resume reviewing when you have time.

File: dashboard/test/models/script_test.rb (hunk 3 above)
Comment (Erin007) [reply dashboard/test/models/script_test.rb:1313]: ah - got it. thanks!

**Review by Erin007 — APPROVED** [review]

File: apps/test/unit/code-studio/components/lessonExtrasTestHelpers.js [hunk 4]
```diff
@@ -70,39 +52,32 @@ export const bonusLevels = [
...
-        type: 'Maze',
-        map: [
-          [0, 0, 0, 0, 0, 0, 0, 0],
-          [0, 1, 1, 1, 4, 1, 3, 0],
-          [0, 1, 1, 0, 1, 1, 4, 0],
-          [0, 1, 0, 1, 1, 0, 1, 0],
-          [0, 1, 1, 1, 0, 1, 1, 0],
-          [0, 1, 1, 0, 1, 1, 1, 0],
-          [0, 2, 1, 1, 1, 1, 1, 0],
-          [0, 0, 0, 0, 0, 0, 0, 0]
```
Comment (davidsbailey) [inline apps/test/unit/code-studio/components/lessonExtrasTestHelpers.js:92]: are we losing test coverage for the lesson extras level to generate a thumbnail from the maze level contents?

**Review by davidsbailey — APPROVED** [review]
LGTM, just one question inline.

File: apps/test/unit/code-studio/components/lessonExtrasTestHelpers.js (hunk 4 above)
Comment (dmcavoy) [reply apps/test/unit/code-studio/components/lessonExtrasTestHelpers.js:92]: I removed this entirely (outside of tests too)  in favor of using thumbnail urls which the levelbuilder sets or the solution image for the level. Thoughts on that decision?

Comment (davidsbailey) [reply apps/test/unit/code-studio/components/lessonExtrasTestHelpers.js:92]: my concern would be if there are some maze levels where we are relying on the existing implementation, that would lose thumbnails as a result of this change. is there any way to confirm whether this is the case?

Comment (dmcavoy) [reply apps/test/unit/code-studio/components/lessonExtrasTestHelpers.js:92]: I'll look into it.

**Comment by dmcavoy** [comment]
Working on some improvements but re-implemented the ability to scroll back to past lesson extras and maze level images work.
lesson-extras-update

(For some reason the gif did not always pick up the whole white box)

**Comment by davidsbailey** [comment]
whoa, I had no idea that's what those arrows did! thanks Dani, please let us know when you're ready for another look.

File: apps/src/templates/progress/LessonExtrasProgressBubble.jsx [hunk 5]
```diff
@@ -1,39 +1,19 @@
 import React, {Component} from 'react';
 import PropTypes from 'prop-types';
-import Radium from 'radium';
 import _ from 'lodash';
 import i18n from '@cdo/locale';
 import TooltipWithIcon from './TooltipWithIcon';
 import {currentLocation} from '../../utils';
-import color from '@cdo/apps/util/color';
+import LessonExtrasFlagIcon from './LessonExtrasFlagIcon';
```
Comment (davidsbailey) [inline apps/src/templates/progress/LessonExtrasProgressBubble.jsx:7]: nice extraction!

File: apps/src/code-studio/components/SublevelCard.jsx [hunk 6]
```diff
@@ -88,8 +89,58 @@ export default class SublevelCard extends React.Component {
...
+    if (isLessonExtra) {
+      url = `?id=${sublevel.id}`;
+      if (sectionId && userId) {
+        // Both sectionId and userId are required to link to a student's work on a bonus level.
+        url += `&section_id=${sectionId}&user_id=${userId}`;
+      } else if (sectionId) {
+        url += `&section_id=${sectionId}`;
+      }
+    } else {
+      url = sublevel.url + location.search;
```
Comment (davidsbailey) [inline apps/src/code-studio/components/SublevelCard.jsx]: please construct the query params using an existing helper like this one: https://github.com/code-dot-org/code-dot-org/blob/a98a0b5bc5dd10e0d6323757c14c591b50277e0d/apps/src/templates/Certificate.jsx#L124

**Review by davidsbailey — COMMENTED** [review]
running to a meeting, but wanted to share comments so far

**Review by davidsbailey — APPROVED** [review]
Looks good, Dani! 

For the record, where did we end up with maze thumbnails -- are we still generating those in code, or are they all static images now?

**Comment by dmcavoy** [comment]
> Looks good, Dani!
> 
> For the record, where did we end up with maze thumbnails -- are we still generating those in code, or are they all static images now?

@davidsbailey they are getting generated in the code.

**Pull request merged**